x-pack/libbeat/management/generate.go

```diff
@@ -0,0 +1,270 @@
+		processors = append(processors, inputId)
+	}
+
+	procInputDataset := DefaultDatasetName
+	if testStreamDataset := streamExpected.GetDataStream().GetDataset(); testStreamDataset != "" {
+		procInputDataset = testStreamDataset
+	}
+
+	//2. Actually add the processors
+	// namespace
+	datastream := generateAddFieldsProcessor(mapstr.M{"dataset": procInputDataset,
+		"namespace": procInputNamespace, "type": procInputType}, "data_stream")
+	processors = append(processors, datastream)
+
+	// dataset
+	event := generateAddFieldsProcessor(mapstr.M{"dataset": procInputDataset}, "event")
+	processors = append(processors, event)
+
+	// source stream
+	streamID := streamExpected.GetId()
+	sourceStream := generateAddFieldsProcessor(mapstr.M{"stream_id": streamID}, "@metadata")
+	processors = append(processors, sourceStream)
+
+	// figure out if we have any existing processors
+	currentProcs, ok := stream["processors"]
+	if !ok {
+		stream["processors"] = processors
+	} else {
+		currentProcsList, ok := currentProcs.([]interface{})
+		if !ok {
+			return nil, fmt.Errorf("error creating list of existing processors, got: %#v", currentProcs)
+		}
+		stream["processors"] = append(processors, currentProcsList...)
+
+	}
+
+	return stream, nil
+}
+
+// ===========
+// Config Processors
+// ===========
+
+func generateAddFieldsProcessor(fields mapstr.M, target string) mapstr.M {
+	return mapstr.M{
+		"add_fields": mapstr.M{
+			"fields": fields,
+			"target": target,
+		},
+	}
+}
+
+// This generates an opaque config blob used by all the beats
+// This has to handle both universal config changes and changes specific to the beats
+// This is a replacement for the AST code that lived in V1
+func generateBeatConfig(unitRaw *proto.UnitExpectedConfig, agentInfo *client.AgentInfo) ([]*reload.ConfigWithMeta, error) {
+	// We aren't guaranteed a DataStream field from the config
+	if unitRaw.GetDataStream() == nil {
+		unitRaw.DataStream = &proto.DataStream{
+			Namespace: DefaultNamespaceName,
+			Dataset:   DefaultDatasetName,
+		}
+	} else {
+		if unitRaw.GetDataStream().GetNamespace() == "" {
+			unitRaw.DataStream.Namespace = DefaultNamespaceName
+		}
+		if unitRaw.GetDataStream().GetDataset() == "" {
+			unitRaw.DataStream.Dataset = DefaultDatasetName
+		}
+	}
+
+	// Generate the config that's unique to a beat
+	metaConfig, err := ConfigTransform.Transform(unitRaw, agentInfo)
+	if err != nil {
+		return nil, fmt.Errorf("error transforming config for beats: %w", err)
+	}
+	return metaConfig, nil
+}
+
+// generate the output config, including shuffling around the `type` key
+// In V1, this was done by the groupByOutputs function buried in the AST init
+func groupByOutputs(outCfg *proto.UnitExpectedConfig) (*reload.ConfigWithMeta, error) {
+	// We still need to emulate the InjectHeadersRule AST code,
+	// I don't think we can get the `Headers()` data reported by the AgentInfo()
+	sourceMap := outCfg.Source.AsMap()
```

Comment:
```suggestion
	sourceMap := outCfg.GetSource().AsMap()
```

Also looking at the implementation of `AsMap()` it will return an allocated by empty map if `Source` is `nil`. That might be something you need to check for, but it at least won't result in `nil` deferences.